### ISSUE TITLE
Configuring Jethro to use SMSbroadcast

### ISSUE THREAD

**Issue by NeilAtw** [tbar0970/jethro-pmm#358]
I could really use some help setting up smsbroadcast in Jethro...

Here's what I have in the configuration section in Admin:

Sms http url: https://api.smsbroadcast.com.au/api-adv.php 
Sms http header template: blank
Sms http post template: username=safeministry&password=****************&to=_RECIPIENTS_COMMAS_&from=_USER_MOBILE_&maxsplit=' . floor(SMS_MAX_LENGTH/160) . '&message=_MESSAGE_ 
Sms Recipient Array parameter: Don’t know what to put here…
Sms Http Response Ok Regex:  blank - don’t know what to put here…
Sms Http Response Error Regex:  blank - don’t know what to put here…
Sms Local Prefix: 0
Sms International Prefix: 61 

Any pointers here would be greatly appreciated!

Thanks,

Neil A.

**Comment by tbar0970** [comment]
@vanoudt  can you offer any wisdom here?

**Comment by tbar0970** [comment]
From an old config file:

// * For smsbroadcast.com.au
//   define('SMS_HTTP_RESPONSE_OK_REGEX', '[\s\S]*_RECIPIENT_INTERNATIONAL_\",\"id\"\:\d*,\"status\"\:\d*,\"status_text\"\:\"Sen');

leaving everything else blank *should* be OK as far as I can tell

**Comment by NeilAtw** [comment]
Thanks Tom… Will give it a try 😊


Rev Neil Atwood (JP)
EA to Bishop Ivan Lee

Phone: (02) 9265 1547
Mob: 0414 777 841
email<mailto:nta@sydney.anglican.asn.au> |  twitter<http://twitter.com/NeilAtw> |  web<http://west.anglicanchurch.net.au> |

PO Box Q190, QVB Post Office, NSW 1230
[http://anglicanchurch.net.au/Anglican_Crest_1.png]


Anglican Diocese of Sydney
PRIVATE AND CONFIDENTIAL You must not disclose, copy or rely on any part of this correspondence if you are not the intended recipient.

**Comment by vanoudt** [comment]
Hi Neil, Tom
Sorry it's taken so long to reply!
Here's what I use at the moment (with passwords/usernames redacted, of course!) (also from an old config file!)

define('SMS_MAX_LENGTH', 320);
define('SMS_COUNTRY_CODE', 61);
define('SMS_HTTP_HEADER_TEMPLATE', "");
define('SMS_HTTP_URL', 'https://api.smsbroadcast.com.au/api-adv.php');
define('SMS_HTTP_POST_TEMPLATE', 'username=USERNAME&password=PASSWORD&to=_RECIPIENTS_COMMAS_&from=_USER_MOBILE_&maxsplit=' . floor(SMS_MAX_LENGTH/160) . '&message=_MESSAGE_');
define('SMS_HTTP_RESPONSE_OK_REGEX', 'OK:[\s]*\+?_RECIPIENT_');

**Comment by vanoudt** [comment]
Actually, scratch that - that isn't the latest config I'm actually using. This is:

`// The maximum SMS length users are allowed to send.  160 chars is usually a one-part SMS.
define('SMS_MAX_LENGTH', 320);

// The two digit country code for mobile phones. Some providers, such as smsbroadcast, use this
define('SMS_COUNTRY_CODE', 61);

// Header template for SMS server api authentication.
// For example, for 5centsms.com.au something like
// "User: yourfivecentregistered@email\r\nApi-Key: YourAPIKey\r\n";
// Note that if you are using \r\n to put in new lines, you should wrap this in ""
//define('SMS_HTTP_HEADER_TEMPLATE',"User: vanoudt@gmail.com\r\nApi-Key: 9HfFCmvyetBzmZZd3GKeuQiHUCZRQb\r\n");
define('SMS_HTTP_HEADER_TEMPLATE', "");

// URL of the server to send SMSes through
//define('SMS_HTTP_URL', 'https://www.5centsms.com.au:/api/v4/sms'); // eg http://www.5centsms.com.au/api/send.php
define('SMS_HTTP_URL', 'https://api.smsbroadcast.com.au/api-adv.php');

define('SMS_LOCAL_PREFIX', '0');
define('SMS_INTERNATIONAL_PREFIX', '61');

// The format for the send-sms POST request
// Can contain keywords _USER_MOBILE_ _USER_EMAIL_ _MESSAGE_ _RECIPIENTS_COMMAS_ _RECIPIENTS_NEWLINES_
// eg 'username=abc&password=xyz&to=_RECIPIENTS_COMMAS_&sender=_USER_MOBILE_&message=_MESSAGE_');
//define('SMS_HTTP_POST_TEMPLATE', 'sender=_USER_MOBILE_&to=_RECIPIENTS_COMMAS_&message=_MESSAGE_');
define('SMS_HTTP_POST_TEMPLATE', 'username=USERNAME&password=PASSWORD&to=_RECIPIENTS_COMMAS_&from=_USER_MOBILE_&maxsplit=' . floor(SMS_MAX_LENGTH/160) . '&message=_MESSAGE_');

// Regex to use to detect success message from the SMS server for each recipient
// _RECIPIENT_ keyword is available.  Leave blank to ignore response.
// eg ^1\|_RECIPIENT_\|[0-9]+\|OK<br>
define('SMS_HTTP_RESPONSE_OK_REGEX', 'OK:[\s]*\+?_RECIPIENT_INTERNATIONAL_');
//define('SMS_HTTP_RESPONSE_OK_REGEX', '[\s\S]*_RECIPIENT_\",\"id\"\:\d*,\"status\"\:\d*,\"status_text\"\:\"Sen');

define('SMS_SAVE_TO_NOTE_BY_DEFAULT', true);

define('SMS_SAVE_TO_NOTE_SUBJECT', 'SMS');

`

**Comment by NeilAtw** [comment]
Thanks Nicholas - really appreciate that.
Will put it to the test very soon.

Cheers
 --

Rev Neil Atwood (JP)
Executive Assistant to Bishop Ivan Lee
Anglican Church Diocese of Sydney

Tel: 02 9265 1547
Mob: 0414 777 841
Email: nta@sydney.anglican.asn.au

PRIVATE AND CONFIDENTIAL
You must not disclose, copy or rely on any part of this correspondence if you are not the intended recipient.

On 3 May 2017 5:12:34 PM Nicholas van Oudtshoorn <notifications@github.com> wrote:

Actually, scratch that - that isn't the latest config I'm actually using. This is:

`// The maximum SMS length users are allowed to send. 160 chars is usually a one-part SMS.
define('SMS_MAX_LENGTH', 320);

// The two digit country code for mobile phones. Some providers, such as smsbroadcast, use this
define('SMS_COUNTRY_CODE', 61);

// Header template for SMS server api authentication.
// For example, for 5centsms.com.au something like
// "User: yourfivecentregistered@email\r\nApi-Key: YourAPIKey\r\n";
// Note that if you are using \r\n to put in new lines, you should wrap this in ""
//define('SMS_HTTP_HEADER_TEMPLATE',"User: vanoudt@gmail.com<mailto:vanoudt@gmail.com>\r\nApi-Key: 9HfFCmvyetBzmZZd3GKeuQiHUCZRQb\r\n");
define('SMS_HTTP_HEADER_TEMPLATE', "");

// URL of the server to send SMSes through
//define('SMS_HTTP_URL', 'https://www.5centsms.com.au:/api/v4/sms'); // eg http://www.5centsms.com.au/api/send.php<http://webdefence.global.blackspider.com/urlwrap/?q=AXicFcxBDsIgEEbh_wpeZGi0pupK48JF3XmCCRBLhCkBLGHhTT2MdfFWX_I2N4x34MtA8q07viinhQI7r2cpafak54Cl3w3Xx8V0_XY4DJDC59yM2EYsT-80C3Fee2MqJZ6UqrXSXlspOeT_YSXF0alsxVCcIoDPCPwABPIotw&Z>
define('SMS_HTTP_URL', 'https://api.smsbroadcast.com.au/api-adv.php');

define('SMS_LOCAL_PREFIX', '0');
define('SMS_INTERNATIONAL_PREFIX', '61');

// The format for the send-sms POST request
// Can contain keywords USER_MOBILE USER_EMAIL MESSAGE RECIPIENTS_COMMAS RECIPIENTS_NEWLINES
// eg 'username=abc&password=xyz&to=RECIPIENTS_COMMAS&sender=USER_MOBILE&message=MESSAGE');
//define('SMS_HTTP_POST_TEMPLATE', 'sender=USER_MOBILE&to=RECIPIENTS_COMMAS&message=MESSAGE');
define('SMS_HTTP_POST_TEMPLATE', 'username=USERNAME&password=PASSWORD&to=RECIPIENTS_COMMAS&from=USER_MOBILE&maxsplit=' . floor(SMS_MAX_LENGTH/160) . '&message=MESSAGE');

// Regex to use to detect success message from the SMS server for each recipient
// RECIPIENT keyword is available. Leave blank to ignore response.
// eg ^1|RECIPIENT|[0-9]+|OK

define('SMS_HTTP_RESPONSE_OK_REGEX', 'OK:[\s]+?RECIPIENT_INTERNATIONAL');
//define('SMS_HTTP_RESPONSE_OK_REGEX', '[\s\S]RECIPIENT","id":\d*,"status":\d*,"status_text":"Sen');

define('SMS_SAVE_TO_NOTE_BY_DEFAULT', true);

define('SMS_SAVE_TO_NOTE_SUBJECT', 'SMS');

`

-

**Comment by tbar0970** [comment]
Note that some bits of the above will only work in the conf.php file, not via the web interface.  But you could modify it to work via the web config like this:
```
SMS_HTTP_POST_TEMPLATE:
username=USERNAME&password=PASSWORD&to=_RECIPIENTS_COMMAS_&from=_USER_MOBILE_&maxsplit=4&message=_MESSAGE_
```
(maxsplit is (I believe) the maximum number of concatenated messages, so you could change it to 1 if you want to ensure each message is only 1 part)

**Comment by NeilAtw** [comment]
Nicholas/Tom,
Thanks again for your assistance with this...
So I decided to configure the whole thing from the conf.php file, and copy/paste your working config into my file (obviously changing username/password).

I'm getting successful sends, but when sending an SMS manually from a persons record in Jethro, I get a red error msg in the sending window (even though the send worked). 
If I remark out the Response regex line, I don't get that error.
Is that OK to leave that way, or is there a better line of code for the Response line?

Many thanks!

Neil

**Comment by tbar0970** [comment]
If you don't have a success regex set up, Jethro can't tell if the send succeeded, so it will just assume the best.  That's OK, but your users might be given a false sense of security in the case of bad numbers etc.

@vanoudt can you offer any more advice on the success regex for smsbroadcast?

**Comment by NeilAtw** [comment]
That's what I assumed Tom... and agree, it would be better to get this sorted for a higher level of confidence in sending.
I can see that smsbroadcast supply a set of responses in their docs (but I have no idea how to "Regx" them!):

Output Parameters: 

Status:
Will show if your messages have been accepted by the API. There are 3 
possible results:
  OK : This message was accepted.
  BAD: This message was invalid. (eg, invalid phone number)
  ERROR: The request failed. (eg, wrong username/password or 
missing a required parameter)

Receiving Number:
The receiving mobile number. This will be shown in international format 
(614xxxxxxxx) regardless of the format it was submitted in.
If you submit an invalid mobile number, the invalid number will be shown 
in the same format as it was submitted.

SMS Reference Number or Error Message:
Will display our reference number for the SMS message, or the reason for 
a failed SMS message.


The API will return 1 line of data for each SMS message you submit. The data is separated by the 
colon character. 
Sample API Output: 
The output from the previous example would be: 
OK: 61400111222:2942263 
OK: 61400222333:2942264 
If you submit an invalid mobile number, the API would return: 
BAD:0400abc111:Invalid Number

**Comment by tbar0970** [comment]
Those docs suggest that Nick's regex above:

`define('SMS_HTTP_RESPONSE_OK_REGEX', 'OK:[\s]+?RECIPIENT_INTERNATIONAL');`

should work OK.

Are you sure you included the international prefix and local prefix in your conf file?

```
define('SMS_LOCAL_PREFIX', '0');
define('SMS_INTERNATIONAL_PREFIX', '61');
```

(this allows Jethro to convert 0417123456 into 61417123456 which is the format that smsbroadcast will return it as)

**Comment by NeilAtw** [comment]
Thanks Tom...
I'm double checking syntax, etc to make sure nothing weird has crept in. ;-)

**Comment by vanoudt** [comment]
Turns out there was also an error in the error handling - just put a pull request through to Tom to fix that. I also struggled upon upgrading... turns out I had left the "u" out from "username"!

**Comment by NeilAtw** [comment]
Ah! Good to hear a solution is coming :)
Thanks Nicholas...

**Comment by tbar0970** [comment]
Fix coming in v2.20 in the next few days.

**Comment by alwr** [comment]
Just a note, on the configuration page (not the conf file) I had to change the regex OK to "OK:[\s]*?_RECIPIENT_INTERNATIONAL_" to get it to work.  I'm not sure it's 100% correct, but it's working!
